File: codes_trial/swift/Array.includes.swift
import Foundation

typealias MyObject = [String: Any?]
typealias MyArray = [Any?]

func prettyArrayOfPrimitives(_ anArrayOfPrimitives: MyArray) -> String {
    var result = "["
    for (arrayItemIndex, arrayItem) in anArrayOfPrimitives.enumerated() {
        if (((arrayItem is String) == false) && ((arrayItem is NSNumber) == false)) {
            continue
        }
        if let arrayItem = arrayItem as? String {
            result += "\"\(arrayItem)\""
        }
        if let arrayItem = arrayItem as? NSNumber {
            result += "\(arrayItem)"
        }
        if ((arrayItemIndex + 1) != anArrayOfPrimitives.count) {
            result += ", "
        }
        
    }
    result += "]"
    return result
}

print("\n// JavaScript-like Array.includes() in Swift")

// There's no JavaScript-like Array.includes() in Swift.
// But, we can use Swift Array.includes() built-in method "Array.contains".

let myFriends: MyArray = ["Alisa", "Trivia"]
print("myFriends: \(prettyArrayOfPrimitives(myFriends))")

var name: String
var isMyFriend: Bool

print("// using Swift Array.includes() built-in method \"Array.contains\"")

name = "Alisa"
print("name: \"\(name)\"")
isMyFriend = myFriends.contains { (myFriend: Any?) -> Bool in
    if let result = myFriend as? String {
        return result == name
    }
    return false
}
print("is my friends includes \"\(name)\": \(isMyFriend)")
// is my friends includes "Alisa": true

name = "Trivia"
print("name: \"\(name)\"")
isMyFriend = myFriends.contains { (myFriend: Any?) -> Bool in
    if let result = myFriend as? String {
        return result == name
    }
    return false
}
print("is my friends includes \"\(name)\": \(isMyFriend)")
// is my friends includes "Trivia": true

name = "Tony"
print("name: \"\(name)\"")
isMyFriend = myFriends.contains { (myFriend: Any?) -> Bool in
    if let result = myFriend as? String {
        return result == name
    }
    return false
}
print("is my friends includes \"\(name)\": \(isMyFriend)")
// is my friends includes "Tony": false

name = "Ezekiel"
print("name: \"\(name)\"")
isMyFriend = myFriends.contains { (myFriend: Any?) -> Bool in
    if let result = myFriend as? String {
        return result == name
    }
    return false
}
print("is my friends includes \"\(name)\": \(isMyFriend)")
// is my friends includes "Ezekiel": false

File: codes/visual_basic_dotnet/array_includes.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.includes() in Visual Basic (.NET)")

        Dim MyFriends As List(Of Object) = New List(Of Object) From {"Alisa", "Trivia"}
        Console.WriteLine($"MyFriends: {JsonStringify(MyFriends)}")

        Dim AName As String
        Dim IsMyFriend As Boolean

        Console.WriteLine("' using Visual Basic (.NET) Array.includes() built-in method ""Array.Contains()""")

        AName = "Alisa"
        IsMyFriend = MyFriends.Contains(AName)
        Console.WriteLine($"is my friends includes ""{AName}"": {IsMyFriend}")
        ' is my friends includes "Alisa": True

        AName = "Trivia"
        IsMyFriend = MyFriends.Contains(AName)
        Console.WriteLine($"is my friends includes ""{AName}"": {IsMyFriend}")
        ' is my friends includes "Trivia": True

        AName = "Tony"
        IsMyFriend = MyFriends.Contains(AName)
        Console.WriteLine($"is my friends includes ""{AName}"": {IsMyFriend}")
        ' is my friends includes "Tony": False

        AName = "Ezekiel"
        IsMyFriend = MyFriends.Contains(AName)
        Console.WriteLine($"is my friends includes ""{AName}"": {IsMyFriend}")
        ' is my friends includes "Ezekiel": False
    End Sub
End Module
